Refactor watch

supcik opened the pull request:
This patch can be used to address the issue #17 (Watch(Event) should be able to unmarshal to user-defined struct). This is what I did:
1. **Simplify** the extraction of the data from the body of the server's reply. I didn't like the idea of having to "glue" lines together and then split them right afterwards. I also wanted something **more robust** that checks if the lines are valid (e.g. a response of exactly 2 lines, the first must start with "event:" and the second line must start with "data:"). I used "scanner.ReadString" to read string of any length.
2. Keep a copy of the json from the second line ("data: <json>") in a private field (Event.payload)
3. Implement "func (e Event) Value(v interface{}) (path string, err error)" that can be user to **unmarshal user-defined struct from the event's payload**
4. Implement a test case for the new method (TestWatchValue).

The advantage of this method is that the API is compatible with the current version.
The drawback is that we have to keep 2 copies of the data; one in the generic form (map[string]interface{}) and another as the original json string.

I also have some additional remarks to the "firego" library:
- I don't think that the "watchMtx" mutex is needed. Indeed you access shared resources, but assigning and reading a boolean is already atomic.
- In the comment of the "Watch" you write "_Only one connection can be established at a time._" I understand that with the current implementation (with a shared ".isWatching") , this is required. But why did you choose this implementation? Is there no use case for several concurrent "watch"?

-- Jacques.

supcik committed:
Add Event.Value method

The Event.Value method can unmarshal values from watch event.
The code for reading the body is also more robust.

supcik committed:
Add ignored files for IntelliJ IDEA

zabawaba99 commented:
Can you run [gofmt](https://golang.org/cmd/gofmt/) against the changes so that the pipeline can run the tests as well?

supcik committed:
apply go fmt

supcik commented:
OK Done.

█  Jacques

On 29 July 2016 at 21:10, Steven Berlanga notifications@github.com wrote:

> Can you run [gofmt](https://golang.org/cmd/gofmt/) against the changes so
> that the pipeline can run the tests as well?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50#issuecomment-236268209, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iEX4FYnYE-Q1iWpT81zGxxvJzOgrks5qalAdgaJpZM4JYHTS
> .

zabawaba99 commented on watch.go:
```diff
@@ -103,48 +121,54 @@ func (fb *Firebase) Watch(notifications chan Event) error {
-			// so we have to add it to our result buffer.
-			evt, isPrefix, scanErr = scanner.ReadLine()
+			payload := make([]string, 2)
+			scanErr = func(payload []string) error {
```
The logic inside of the closure is a little confusing. I like the idea of enforcing that we should only ever get two lines from the connection. What about something like

``` go
var (
    lineCount int
    line      string
)
line, scanErr = scanner.ReadString('\n')
for ; lineCount != len(payload); line, scanErr = scanner.ReadString('\n') {
    if scanErr != nil {
        break scanning
    }

    line = strings.Trim(line, " \r\n")
    if len(line) == 0 { // empty line
        scanErr = errors.New("Bad formated body")
        break scanning
    }
    payload[lineCount] = line
    lineCount++
}
```

supcik replied:
It is indeed an option. The closure is for me a way to avoid the "break"
with a label. which I don't really like; but this is probably a question of
personal style.

The other benefits of my solution is that it handles correctly the case
where a reply has more than 2 lines. Perhaps this is what causes the
confusion.

█  Jacques

On 31 July 2016 at 20:54, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r72908852:
> 
> > @@ -103,48 +121,54 @@ func (fb *Firebase) Watch(notifications chan Event) error {
> >             //      event: put
> >             //      data: {"path":"/","data":{"foo":"bar"}}
> > -           var evt []byte
> > -           var dat []byte
> > -           isPrefix := true
> > 
> > ## \-           var result []byte
> > -           // For possible results larger than 64 \* 1024 bytes (MaxTokenSize)
> > -           // we need bufio#ReadLine()
> > -           // 1. step: scan for the 'event:' part. ReadLine() oes not return the \n
> > -           // so we have to add it to our result buffer.
> > -           evt, isPrefix, scanErr = scanner.ReadLine()
> > -           payload := make([]string, 2)
> > -           scanErr = func(payload []string) error {
> 
> The logic inside of the closure is a little confusing. I like the idea of
> enforcing that we should only ever get two lines from the connection. What
> about something like
> 
> var (
>     lineCount int
>     line      string
> )
> line, scanErr = scanner.ReadString('\n')for ; lineCount != len(payload); line, scanErr = scanner.ReadString('\n') {
>     if scanErr != nil {
>         break scanning
>     }
> 
> ```
> line = strings.Trim(line, " \r\n")
> if len(line) == 0 { // empty line
>     scanErr = errors.New("Bad formated body")
>     break scanning
> }
> payload[lineCount] = line
> lineCount++
> ```
> 
> }
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r72908852,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iL5RvdcFrr5J1iIEVGZLhtqFEygQks5qbO94gaJpZM4JYHTS
> .

zabawaba99 replied:
> It is indeed an option. The closure is for me a way to avoid the "break"
> with a label. which I don't really like; but this is probably a question of
> personal style.

If the case is to avoid having to break a label, can you extract this into it's own function then?

> The other benefits of my solution is that it handles correctly the case
> where a reply has more than 2 lines. Perhaps this is what causes the
> confusion.

Would you mind giving me an example of a string that would have more than two lines that would not be handled correctly by both pieces of code?

supcik replied:
It's OK to make a local function. I can do this.

Suppose that the reply is something like this:

event: someEvent
data: { some JSON data }
_other-header: other-info_
<empty line>

I know that this is not supposed to occur, but I wanted to handle such
cases anyhow.

Additionally, as far as I understand your code, you are ignoring the first
line:

line, scanErr = scanner.ReadString('\n') <- first line read here

> for ; lineCount != len(payload); line, scanErr = scanner.ReadString('\n')
> { <- first line overwritten here

but perhaps I missed something in your code.

█  Jacques

On 31 July 2016 at 21:26, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r72909323:
> 
> > @@ -103,48 +121,54 @@ func (fb *Firebase) Watch(notifications chan Event) error {
> >             //      event: put
> >             //      data: {"path":"/","data":{"foo":"bar"}}
> > -           var evt []byte
> > -           var dat []byte
> > -           isPrefix := true
> > 
> > ## \-           var result []byte
> > -           // For possible results larger than 64 \* 1024 bytes (MaxTokenSize)
> > -           // we need bufio#ReadLine()
> > -           // 1. step: scan for the 'event:' part. ReadLine() oes not return the \n
> > -           // so we have to add it to our result buffer.
> > -           evt, isPrefix, scanErr = scanner.ReadLine()
> > -           payload := make([]string, 2)
> > -           scanErr = func(payload []string) error {
> 
> It is indeed an option. The closure is for me a way to avoid the "break"
> with a label. which I don't really like; but this is probably a question of
> personal style.
> 
> If the case is to avoid having to break a label, can you extract this into
> it's own function then?
> 
> The other benefits of my solution is that it handles correctly the case
> where a reply has more than 2 lines. Perhaps this is what causes the
> confusion.
> 
> Would you mind giving me an example of a string that would have more than
> two lines that would not be handled correctly by both pieces of code?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r72909323,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iJZzdirda6_jdfY_3tTU0x47GU8nks5qbPbfgaJpZM4JYHTS
> .

zabawaba99 commented on watch.go:
```diff
@@ -103,48 +121,54 @@ func (fb *Firebase) Watch(notifications chan Event) error {
-			if scanErr != nil {
+
+			var eventType string
+			if strings.HasPrefix(payload[0], "event:") {
```
Can you handle the negative case first?

``` go
if !strings.HasPrefix(payload[0], "event:") {
    scanErr = errors.New("First line does not start with event:")
    break scanning
}

// do the string trimming
```

supcik replied:
No problem. I will do this.

█  Jacques

On 31 July 2016 at 20:56, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r72908889:
> 
> > -               dat, isPrefix, scanErr = scanner.ReadLine()
> > -               if scanErr != nil {
> > -                   break scanning
> > -               }
> > -               result = append(result, dat...)
> > -               if !isPrefix {
> > -                   break
> > -               }
> > -           }
> > -           // Again we add the \n
> > -           result = append(result, '\n')
> > -           _, _, scanErr = scanner.ReadLine()
> > -           if scanErr != nil {
> >   +
> > -           var eventType string
> > -           if strings.HasPrefix(payload[0], "event:") {
> 
> Can you handle the negative case first?
> 
> if !strings.HasPrefix(payload[0], "event:") {
>     scanErr = errors.New("First line does not start with event:")
>     break scanning
> }
> // do the string trimming
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r72908889,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iAUufoxcam9Yu4VVm3r_qNYBGXVkks5qbO_HgaJpZM4JYHTS
> .

zabawaba99 commented on watch.go:
```diff
@@ -103,48 +121,54 @@ func (fb *Firebase) Watch(notifications chan Event) error {
-			txt := string(result)
-			parts := strings.Split(txt, "\n")
+			var eventData string
+			if strings.HasPrefix(payload[1], "data:") {
```
Can you handle the negative case first?

zabawaba99 commented on watch.go:
```diff
@@ -180,7 +204,8 @@ func (fb *Firebase) Watch(notifications chan Event) error {
 			case "rules_debug":
-				log.Printf("Rules-Debug: %s\n", txt)
+				txt, _ := ioutil.ReadAll(resp.Body)
+				log.Printf("Rules-Debug: %s\n", string(txt))
```
Wouldn't this be the `eventData`?

supcik replied:
I don't know how the "rules_debug" is used. As far as I remember, your
original code gives the full body, including the "event:" line. In case of
malformed response, the eventData would be empty. So this is why I chose to
re-read the full body.

█  Jacques

On 31 July 2016 at 20:57, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r72908920:
> 
> > @@ -180,7 +204,8 @@ func (fb *Firebase) Watch(notifications chan Event) error {
> > 
> > ```
> >             // TODO: handle
> >         case "rules_debug":
> > ```
> > -               log.Printf("Rules-Debug: %s\n", txt)
> > -               txt, _ := ioutil.ReadAll(resp.Body)
> > -               log.Printf("Rules-Debug: %s\n", string(txt))
> 
> Wouldn't this be the eventData?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r72908920,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iPTtex2wypPuhQlU-SfjUBIVJKpsks5qbPAZgaJpZM4JYHTS
> .

zabawaba99 replied:
I'll go ahead and run a test for this. My fear with reading the entire response body is that it will read until it get's an `EOF` which may cause other messages to be missed.

supcik replied:
You are right. It should only read until an empty line. I will fix it.

█  Jacques

On 31 July 2016 at 21:21, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r72909276:
> 
> > @@ -180,7 +204,8 @@ func (fb *Firebase) Watch(notifications chan Event) error {
> > 
> > ```
> >             // TODO: handle
> >         case "rules_debug":
> > ```
> > -               log.Printf("Rules-Debug: %s\n", txt)
> > -               txt, _ := ioutil.ReadAll(resp.Body)
> > -               log.Printf("Rules-Debug: %s\n", string(txt))
> 
> I'll go ahead and run a test for this. My fear with reading the entire
> response body is that it will read until it get's an EOF which may cause
> other messages to be missed.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r72909276,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iItELjgeYLWwDhws0ZwpOvgwld0Rks5qbPXVgaJpZM4JYHTS
> .

zabawaba99 commented on watch.go:
```diff
@@ -21,7 +23,23 @@ type Event struct {
+	payload string
+}
+
+func (e Event) Value(v interface{}) (path string, err error) {
```
Why would it matter to return the path here? It's already available through the `Event` struct.

zabawaba99 replied:
I apologize that it took so long for me to response (life got a little crazy). In this case I think it would be worth changing the return type to just be `error` and continue to let uses access the path via the property. In the future, if the api is changed, then we could potentially double back and change this.

supcik commented on watch.go:
```diff
@@ -21,7 +23,23 @@ type Event struct {
+	payload string
+}
+
+func (e Event) Value(v interface{}) (path string, err error) {
```
If you decide to deprecate the current solution (with the
map[string]interface{} style), then you can save CPU time and RAM usage.
But then, the path will not be extracted and will only be available through
the "Value" method.

█  Jacques

On 31 July 2016 at 21:00, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r72908982:
> 
> > @@ -21,7 +23,23 @@ type Event struct {
> >     // Path to the data that changed
> >     Path string
> >     // Data that changed
> > -   Data interface{}
> > -   Data    interface{}
> > -   payload string
> >   +}
> >   +
> >   +func (e Event) Value(v interface{}) (path string, err error) {
> 
> Why would it matter to return the path here? It's already available
> through the Event struct.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r72908982,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iEpgNCe7MkwONavjYuGrq89V_e6Qks5qbPDAgaJpZM4JYHTS
> .

zabawaba99 replied:
In order for us to see those gains though, we would have to remove the old style. For the time being, I think we should return an error from this function and just have the consumer access the path through the `Event` struct.

zabawaba99 replied:
Also, could you add documentation for this function?

supcik replied:
This is up to you. If you want to keep the "old style" forever, then I
agree that it is OK to return the error only. But if you plan to make the
system more efficient by deprecating the "old style", then you will need a
way to get the path and you will probably have to then change the API.

So I suggest that you keep the path in the return values, but it is your
package and you can decide what you want. Just let me know your final
decision and I will code accordingly.

█  Jacques

On 2 August 2016 at 04:29, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r73085472:
> 
> > @@ -21,7 +23,23 @@ type Event struct {
> >     // Path to the data that changed
> >     Path string
> >     // Data that changed
> > -   Data interface{}
> > -   Data    interface{}
> > -   payload string
> >   +}
> >   +
> >   +func (e Event) Value(v interface{}) (path string, err error) {
> 
> In order for us to see those gains though, we would have to remove the old
> style. For the time being, I think we should return an error from this
> function and just have the consumer access the path through the Event
> struct.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r73085472,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iG6KYx7CzW6h4pMTrynlV2PPcjjnks5qbqt6gaJpZM4JYHTS
> .

supcik replied:
Sure. I will do it as soon as the question concerning the path as a return
value is closed.

█  Jacques

On 2 August 2016 at 04:29, Steven Berlanga notifications@github.com wrote:

> In watch.go
> https://github.com/zabawaba99/firego/pull/50#discussion_r73085481:
> 
> > @@ -21,7 +23,23 @@ type Event struct {
> >     // Path to the data that changed
> >     Path string
> >     // Data that changed
> > -   Data interface{}
> > -   Data    interface{}
> > -   payload string
> >   +}
> >   +
> >   +func (e Event) Value(v interface{}) (path string, err error) {
> 
> Also, could you add documentation for this function?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/zabawaba99/firego/pull/50/files/91637d68c23bcd6761930b23882b18621b759365#r73085481,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAB2iJW2_yBnFy_3qB_6NX5vnUo0-kpiks5qbquGgaJpZM4JYHTS
> .

supcik committed:
Changes according to original author's comments

supcik committed:
Fix according to original author's comment

rules_debug is not part to the Firebase Database REST API. Changes done according to author's comment.

supcik committed:
Bug fix
